algebird-core/src/main/scala/com/twitter/algebird/Scan.scala

```diff
@@ -0,0 +1,361 @@
+      destructor(stateBeforeProcessingI)
+  }
+
+  /**
+    * A Scan that returns the number N for the Nth input (starting from 0)
+    */
+  val index: Aux[Any, Long, Long] = fromStreamLike(0L)(n => (n, n+1))
+
+  def identity[A] = fromFunction[A, A](x => x)
+
+  /**
+   *
+   * @param initStateCreator A call-by-name method that allocates new mutable state
+   * @param presentAndUpdateStateFn A function that both presents the output value, and has the side-effect of updating the mutable state
+   * @tparam I
+   * @tparam S
+   * @tparam O
+   * @return A Scan that safely encapsulates state while it's doing its thing.
+   */
+  def mutable[I, S, O](initStateCreator: => S)(presentAndUpdateStateFn: (I, S) => O): Aux[I, S, O] =
+    new Scan[I, O] {
+      override type State = S
+      override def initialState = initStateCreator
+      override def presentAndNextState(i: I, s: S): (O, S) = (presentAndUpdateStateFn(i, s), s)
+    }
+
+  /**
+   * The trivial scan that always returns the same value, regardless of input
+   * @param t
+   * @tparam T
+   */
+  def const[T](t: T): Aux[Any, Unit, T] = from(()) { (_, _) =>
+    (t, ())
+  }
+
+  /**
+   *
+   * @param aggregator
+   * @param initState
+   * @tparam A
+   * @tparam B
+   * @tparam C
+   * @return A scan which, when given [a_1, ..., a_n] outputs [c_1, ..., c_n] where
+   *         c_i = initState + aggregator.prepare(a_1) + ... + aggregator.prepare(a_i)
+   */
+  def fromAggregator[A, B, C](aggregator: Aggregator[A, B, C], initState: B): Aux[A, B, C] =
+    new Scan[A, C] {
+      override type State = B
+
+      override val initialState = initState
+
+      override def presentAndNextState(a: A, stateBeforeProcessingI: B): (C, B) = {
+        // nb: the order of the arguments to semigroup.plus here is what determines the order of the final summation;
+        // this matters because not all semigroups are commutative
+        val stateAfterProcessingA =
+          aggregator.semigroup.plus(stateBeforeProcessingI, aggregator.prepare(a))
+        (aggregator.present(stateAfterProcessingA), stateAfterProcessingA)
+      }
+    }
+
+  /**
+   *
+   * @param monoidAggregator
+   * @tparam A
+   * @tparam B
+   * @tparam C
+   * @return A scan which, when given [a_1, ..., a_n] outputs [c_1, ..., c_n] where
+   *         c_i = monoidAggregator.monoid.zero + aggregator.prepare(a_1) + ... + aggregator.prepare(a_i)
+   */
+  def fromMonoidAggregator[A, B, C](monoidAggregator: MonoidAggregator[A, B, C]): Aux[A, B, C] =
+    fromAggregator(monoidAggregator, monoidAggregator.monoid.zero)
+
+  /**
+   *
+   * @param aggregator
+   * @param initState
+   * @tparam A
+   * @tparam B
+   * @tparam C
+   * @return  A scan which, when given [a_1, ..., a_n] outputs [c_1, ..., c_n] where
+   *         c_i = aggregator.prepare(a_i) + ... + aggregator.prepare(a_1) + initState
+   */
+  def fromAggregatorReverse[A, B, C](aggregator: Aggregator[A, B, C], initState: B): Aux[A, B, C] =
```

Comment:
to me this seems like it should be a combinator on `Aggregator` or really `Semigroup` and not here. Like `def reverse` method on Aggregator.

Reply:
I'd like to push back on this a little bit. It's not a-priori obvious to me from the signature of this method whether the state is calculated by adding on the right or adding on the left (in essence doing a `foldRight` or a `foldLeft` to calculate the state of each element of the stream). From my perspective, and having both methods available calls out the ambiguity here, forcing the caller to think about which semantics they prefer, assuming they care.

That said, maybe the names `fromAggregator` and `fromAggregatorReverse` show too much bias on my part for the semantics of one side over the other. How would you feel about me renaming them to `fromAggregatorRight` and `fromAggregatorLeft` or something along those lines?

Reply:
My concern is that Aggregator already has order semantics. You may not like that, but it does: order matters on `apply` and `applyOption` on `Aggregator`. So, trying to change that in a related abstraction seems inconsistent to me. I'm concerned that we have already crossed this bridge, and I think it is better to simply be consistent now.

Next: I'm skeptical this API is actually very commonly called in any case. Would a compromise be to just remove it and not make any other changes? Adding two combinators on Semigroup:

```scala
def intercalate[A](sg: Semigroup[A], middle: A): Semigroup[A] =
  new Semigroup[A] {
    def plus(left: A, right: A) = sg.plus(sg.plus(a, middle), b)
  }

def reverse[A](sg: Semigroup[A]): Semigroup[A] =
  new Semigroup[A] {
    def plus(left: A, right: A) = sg.plus(right, left)
  }
```

is something we can do or not (and the implications for Aggregator/MonoidAggregator) in a separate PR.

Reply:
This is a good point, and seems reasonable enough to me.

Irrelevant to this PR, but relevant to the broader discussion: Are you worried about the performance impact of a `reverse` method on `Semigroup`s essentially destroying any custom `SumOption` methods?

Reply:
I am worried about that. `sumOption` composes terribly, but one thing we do on products is materialize 1000 groups of items, then do sumOption after materialization. We could do the same thing with reverse: materialize an group of 1000, call underlying sumOption, repeat.

Reply:
Seems like a reasonable workaround, although now you're materializing 1000 items even for monoids that don't have a custom `sumOption` which might have weird memory impact. But, to your point, `sumOption` composes badly and that's probably the right tradeoff. And, even more to the point, this is all outside the scope of this PR.

In the meantime: I've deleted the `reverse` methods.